packages/devtools-network-console/src/utility/editor-grid-dom-utils.ts

```diff
@@ -0,0 +1,70 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+import { findInAncestors } from './dom';
+
+/**
+ * When deleting a row, we need to traverse to a new grid element or else focus is lost. However,
+ * the problem is that focus management within a functional React component where all of the
+ * individually-rendered DOM elements exist in different components is "hard."
+ *
+ * This function has special knowledge about the internals of EditorGrid and its sub-components.
+ *
+ * It follows this algorithm:
+ *  - If we are on the last row:
+ *   - If there is only one row, focus to the nearest pivot header
+ *   - Otherwise, focus on the "key" field of the previous row
+ *  - If we are not on the last "real" row (a row with data), we focus on the next row "key" field.
+ *  - If we are on the last "real" row, then:
+ *   - If we're also on the first "real" row (meaning there is only one row), we focus to the "key"
+ *     field of the "input" row. If there is no "input" row, we focus to the nearest pivot header.
+ *   - Otherwise we focus to the "key" field of the previous sibling row.
+ *
+ * @param currentTarget The button which is being activated
+ * @param fallbackTarget A fallback target to focus if the algorithm fails
+ */
+export function focusNextGridElement(currentTarget: HTMLButtonElement, fallbackTargetSelector: string): void {
+    const currentRow = findInAncestors(currentTarget, e => e.classList.contains('nc-editor-row')) as HTMLDivElement;
+    const grid = findInAncestors(currentTarget, e => e.classList.contains('nc-editor-grid')) as HTMLDivElement;
+    // invariant: currentRow and grid should be defined
+    if (!currentRow || !grid) {
+        throw new Error('Current target does not appear to be part of an editor grid.');
+    }
+
+    const rows = grid.querySelectorAll('.nc-editor-row') as NodeListOf<HTMLDivElement>;
+    const currentRowIndex = Array.from(rows).indexOf(currentRow);
+    // invariant: currentRowIndex should be > -1
+    if (!rows.length || currentRowIndex === -1) {
+        throw new Error('Current target does not appear to be part of an editor grid or the grid is in an invalid state.');
+    }
+
+    const isOnLastRow = rows.length === currentRowIndex + 1;
+    if (isOnLastRow) {
+        const fallbackTarget = document.querySelector(fallbackTargetSelector) as HTMLElement;
+        scheduleFocus(fallbackTarget);
```

Comment:
nit: I'd expect scheduleFocus to be used to trigger some sort of async update to ensure we don't update focus all the time or something similar. But it looks like we just call target.focus(). should we consider doing that directly here?

Reply:
I'm torn. When I first built that utility, we would do delay-focus() calls and sometimes post a message to the host to assert focus. Now we don't post a message at all, and we don't ever delay-focus. 

I've been hemming and hawing about whether to keep this or not. I'm really not sure.

Reply:
I'm ok to keep this as is. Makes it easier to modify consistently in the future